fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/PlanSerialization.java

```diff
@@ -0,0 +1,232 @@
+/*
+ * PlanSerialization.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2023 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.serialization;
+
+import com.apple.foundationdb.annotation.ProtoMessage;
+import com.apple.foundationdb.record.PlanSerializable;
+import com.apple.foundationdb.record.PlanSerializationContext;
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.RecordMetaDataProto;
+import com.apple.foundationdb.record.metadata.expressions.LiteralKeyExpression;
+import com.google.common.base.Verify;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Sets;
+import com.google.common.collect.Streams;
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.Message;
+import org.apache.commons.lang3.tuple.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.net.URL;
+import java.net.URLConnection;
+import java.nio.charset.StandardCharsets;
+import java.util.Enumeration;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+/**
+ * Class to facilitate serialization and deserialization of
+ * {@link com.apple.foundationdb.record.query.plan.plans.RecordQueryPlan}s.
+ */
+public class PlanSerialization {
+    private static final Logger logger = LoggerFactory.getLogger(PlanSerialization.class);
+
+    private static final String PREFIX = "META-INF/services/";
+
+    private static final Map<Class<? extends Message>, Method> fromProtoMapMethodMap =
+            loadFromProtoMethodMap();
+
+    /**
+     * Parse a single line from the given configuration file, adding the
+     * name on the line to set of names if not already seen.
+     */
+    private static int parseLine(BufferedReader r, int lc, Set<String> namesAlreadySeen, Set<String> names)
```

Comment:
Why are we parsing the config files ourselves? Is this because the `ServiceLoader` returns a list of instantiated instances, and you need a collection of `Class<?>` objects instead? And the `ServiceLoader` can't load the instances because they need a default constructor to be service loadable, which this is not.

I don't love having custom parsing logic of these fields here (or the use of reflection, if I'm honest). I think the way we've typically done this before is by introducing a `Factory` class that can have default constructors, and then making the `Factory` the thing in the service loader.

For example, I think if this were the interface:

```java
public interface PlannerDeserializer<M extends Message, P extends PlannerSerializable> {
    Class<M> getProtoMessageClass();
    P fromProto(ProtoSerializationContext serializationContext, M protoMessage);
}
```

Then I think you could use a `ServiceLoader` to collect those factory implementations without the need for custom parsing logic here or reflection

Reply:
Wouldn't using a `Factory` interface let us avoid that before Java 17 (and avoid reflection post-Java 17)? I think it may also let us avoid the annotation that was added in this PR, if I'm understanding the annotation's purpose correctly

Reply:
So I played with your suggested approach quite a bit to get this as concise as possible. I had thought about that before I went the route you see in the PR but I ended up quite disliking it for the additional boilerplate it introduces. There are 200ish implementations that now have to define a static deserialization class. Quite tedious. I also started to try to get the serializer autogenerated through the `ProtoMessage` annotation but that didn't work as the generated code needed to create the `AutoService` annotation which for some reason didn't get picked up to do its magic in `META-INF`. I will try to get Mike's opinion on this. It looks as if we just need to wait until that bug in the JVM is fixed and I would be able to cut a lot of code.